### PR TITLE
[Calyx] Verify that sources of group ports and assignments are not logically complex.

### LINKED ISSUE #1769: [Calyx] Inlining `%wire` incorrectly assigns guard statements to ports
The following MLIR code:
```mlir
module  {
  calyx.program  {
    calyx.component @main(%in0: i32, %in1: i32, %in2: i32, %clk: i1, %reset: i1, %go: i1) -> (%out0: i32, %done: i1) {
      %true = hw.constant true
      %c0_i32 = hw.constant 0 : i32
      %while_0_arg2_reg.in, %while_0_arg2_reg.write_en, %while_0_arg2_reg.clk, %while_0_arg2_reg.reset, %while_0_arg2_reg.out, %while_0_arg2_reg.done = calyx.register "while_0_arg2_reg" : i32, i1, i1, i1, i32, i1
      %while_0_arg1_reg.in, %while_0_arg1_reg.write_en, %while_0_arg1_reg.clk, %while_0_arg1_reg.reset, %while_0_arg1_reg.out, %while_0_arg1_reg.done = calyx.register "while_0_arg1_reg" : i32, i1, i1, i1, i32, i1
      %while_0_arg0_reg.in, %while_0_arg0_reg.write_en, %while_0_arg0_reg.clk, %while_0_arg0_reg.reset, %while_0_arg0_reg.out, %while_0_arg0_reg.done = calyx.register "while_0_arg0_reg" : i32, i1, i1, i1, i32, i1
      %bb2_arg1_reg.in, %bb2_arg1_reg.write_en, %bb2_arg1_reg.clk, %bb2_arg1_reg.reset, %bb2_arg1_reg.out, %bb2_arg1_reg.done = calyx.register "bb2_arg1_reg" : i32, i1, i1, i1, i32, i1
      %bb2_arg0_reg.in, %bb2_arg0_reg.write_en, %bb2_arg0_reg.clk, %bb2_arg0_reg.reset, %bb2_arg0_reg.out, %bb2_arg0_reg.done = calyx.register "bb2_arg0_reg" : i32, i1, i1, i1, i32, i1
      calyx.wires  {
        calyx.group @assign_while_0_init  {
          calyx.assign %while_0_arg0_reg.in = %in0 : i32
          calyx.assign %while_0_arg0_reg.write_en = %true : i1
          calyx.assign %while_0_arg1_reg.in = %c0_i32 : i32
          calyx.assign %while_0_arg1_reg.write_en = %true : i1
          calyx.assign %while_0_arg2_reg.in = %c0_i32 : i32
          calyx.assign %while_0_arg2_reg.write_en = %true : i1
          %0 = comb.and %while_0_arg0_reg.done, %while_0_arg1_reg.done : i1
          %1 = comb.and %0, %while_0_arg2_reg.done : i1
          calyx.group_done %1 : i1
        }
      }
      calyx.control  {
        calyx.enable @assign_while_0_init
      }
    }
  }
}
```

Generates the following assignment in the Calyx code using `circt-translate --export-calyx`:
```
      assign_while_0_init[done] = ((while_0_arg0_reg.done & while_0_arg1_reg.done) & while_0_arg2_reg.done);
```
This is incorrect. Calyx does not allow for complex guard statements to be assigned to ports. It should instead be:
```
 assign_while_0_init[done] = ((while_0_arg0_reg.done & while_0_arg1_reg.done) & while_0_arg2_reg.done) ? 1'd1;
```

The problem is that that guard inliner for the emitter incorrectly transforms `calyx.group_done %1 : i1` into the above by recursively inlining `%1`.

Two possible solutions are:
1. Always using `calyx.group_done %true, %1 ?` or something so that there is an explicit port
2. Making the emitter less smart and just structurally instantiate the wires into the relevant subcomponent like `std_add` etc.

### PR DESCRIPTION
**Pull request by cgyurgyik**
As discussed in #1769, some verifiers should be added to avoid giving the sources of group ports / assignments logically complex expressions. Also adds a verifier to ensure that `GroupOp` isn't being used as a `CombGroupOp`, and makes the necessary refactoring.

```mlir
import "primitives/std.lib";

component main() -> () {
  cells {
    a = std_reg(2);
  }
  wires {
    group A {
      a.in = 2'd1;
      a.write_en = 1'b1;
      A[done] = (a.done & 1'b1) ? 1'b1;           // <-- 1. Allowed
      // A[done] =  1'b1;                            <-- 2. Not allowed, because then this is just a combinational group.
      // A[done] = 1'b1 ? (a.done & 1'b1);           <-- 3. Not allowed, complex source.
      // A[done] = (a.done & 1'b1) ? (1'b1 & 1'b1);  <-- 4. Not allowed, complex source.
      // A[done] = (a.done & 1'b1);                  <-- 5. Not allowed, complex source.
    }
  }
  control { A; }
}
```

This and f8fcb02 in the SCFCalyx dialect should close #1769. Happy to hear better wording for the errors :-)

### CODE REVIEW

**Comment by rachitnigam** [comment]
One other possible solution is that when a complex expression is present in the source position of an assignment, we structurally instantiate the closest operation as a cell and then make the source of the assignment the output of that cell:

```mlir
%0 = comb.and %while_0_arg0_reg.done, %while_0_arg1_reg.done : i1
%1 = comb.and %0, %while_0_arg2_reg.done : i1
calyx.group_done %1 : i1
```
into:
```mlir
%and.out, %and.left, %and.right = calyx.instance std_and ...
...
%0 = comb.and %while_0_arg0_reg.done, %while_0_arg1_reg.done : i1
calyx.assign %and.left, %0 : i1
calyx.assign %and.right, %while_0_arg2_reg.done : i1
calyx.group_done %and.out : i1
```

**Comment by cgyurgyik** [comment]
So you're saying avoid `comb::{xor, or, and}` and always use Calyx primitives to implement logically complex values.

What benefits do you see to doing this? I ask mostly because this is going to require a lot of refactoring :)

**Comment by rachitnigam** [comment]
Ah, actually, I see the problem. You can't just instantiate the first wire; you have to instantiate all of them instead of inlining.

The only benefit is that you can use wires homogeneously everywhere you can use a port instead of being restricted to using wires in guard positions. Once Calyx supports wires, this will end up being the case.

I'm going to let you decide if it's worth it or not. If it isn't, then we can open an issue saying that once Calyx supports wires, we should start allowing wires in source position again and eliminate this verification check.

**Comment by cgyurgyik** [comment]
> Ah, actually, I see the problem. You can't just instantiate the first wire; you have to instantiate all of them instead of inlining.
> 
> The only benefit is that you can use wires homogeneously everywhere you can use a port instead of being restricted to using wires in guard positions. Once Calyx supports wires, this will end up being the case.
> 
> I'm going to let you decide if it's worth it or not. If it isn't, then we can open an issue saying that once Calyx supports wires, we should start allowing wires in source position again and eliminate this verification check.

Ok. I've simply added a TODO for now. For sake of time, leaving this for now, to ensure that proper Calyx programs are lowered to the native compiler. When wires are supported, we can remove it then.

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 1]
```diff
@@ -76,8 +78,30 @@ SmallVector<Direction> direction::unpackAttribute(Operation *component) {
...
+/// native compiler. Remove this when Calyx supports wire declarations.
+/// See: https://github.com/llvm/circt/pull/1774 for context.
+template <typename Op>
+static LogicalResult verifyNotComplexSource(Op op) {
+  Operation *definingOp = op.src().getDefiningOp();
+  if (definingOp == nullptr)
+    // This is a port of the parent component.
+    return success();
+
+  if (isa<comb::XorOp, comb::AndOp, comb::OrOp>(definingOp))
```
Comment (mikeurbach) [inline lib/Dialect/Calyx/CalyxOps.cpp:96]: Can this just check if the definingOp is in the `comb` dialect?

**Review by mikeurbach — APPROVED** [review]
I think this makes sense based on my understanding of Calyx's semantics.

File: lib/Dialect/Calyx/CalyxOps.cpp (hunk 1 above)
Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp:96]: Sure if you show me how :-). I see Operation* has a getDialect(), but I don't think that's what I want?

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp]: Nevermind, just using `isa` works!

**Comment by cgyurgyik** [comment]
> I think this makes sense based on my understanding of Calyx's semantics.

Thanks for the review. Just to re-iterate, this is a temporary validation so that current Calyx programs will lower to the native compiler. Eventually wires will be supported in both the source and guard positions.

**Pull request merged**